runtime/src/qio/qio_formatted.c

```diff
@@ -361,6 +361,86 @@ qioerr _append_char(char* restrict * restrict buf, size_t* restrict buf_len, siz
   return 0;
 }
 
+qioerr read_entire_file(const int threadsafe,qio_channel_t* restrict ch, const char* restrict* restrict out, int64_t* restrict len_out, ssize_t maxlen)
+{  
+  qioerr err,err_reading_file=0;
+  int64_t file_length;
+  char* ret=NULL;
+  ssize_t amt;
+  err_t errcode;
+  int64_t fileLenGuess=1024;
+  if( threadsafe ) 
+  {
+    err = qio_lock(&ch->lock);
+    if( err ) return err;
+  }
+
+  err = qio_channel_mark(false, ch);
+  if( err ) goto unlock;
+
+  //getting file length
+  err=qio_file_length(ch->file,&file_length);
+  //if file length successfully obtained
+  if(!err)
+  {
+    ret=qio_malloc(file_length+1);
+    ret[0]='\0';
+    err=qio_channel_read(false, ch, ret, file_length, &amt);
+  }
+//  else
+//  {
+//    fileLenGuess=1024;
+//  }
+  //in case file size information is not available or data was added to the file while reading
+  if(err)
+  {
+    ret=qio_malloc(fileLenGuess+1); //allocate an initial buffer size
+    while(true)
+    {
+      ret[0]='\0';
+      err_reading_file = qio_channel_read(false, ch, ret, fileLenGuess, &amt);
```

Comment:
you need to track how much you have already read (which each call will store into amt) and read into `&ret[pos]`. You can't just write `&ret[amt]` because amt is just the amount read from this read, but you need the total from all reads.

Reply:
As it is now, you will be replacing the data in the first portion of `ret` but not filling in the rest